### ISSUE TITLE
AI Based Resume Screening System

### ISSUE THREAD

**Issue by chavandnyanesh2000-byte** [chavandnyanesh2000-byte/Project#1]
import streamlit as st
import pdfplumber
import spacy
import re
import numpy as np
from transformers import AutoTokenizer, AutoModel
import torch
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.cluster import KMeans
import pandas as pd
from io import BytesIO

# Load NLP model
nlp = spacy.load("en_core_web_sm")

# Set device: GPU if available, else CPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Load tokenizer and BERT model on the selected device
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
model = AutoModel.from_pretrained("bert-base-uncased").to(device)

# Function to extract text from PDF
def extract_text_from_pdf(file):
    text = ""
    with pdfplumber.open(file) as pdf:
        for page in pdf.pages:
            text += page.extract_text() or ""
    return text

# Function to parse resume using spaCy
def parse_resume(text):
    doc = nlp(text.lower())
    skills = []
    experience = []
    education = []

    # Extract skills (simple keyword-based, extendable with a skills list)
    skill_keywords = ["python", "java", "sql", "machine learning", "project management", "data analysis"]
    for token in doc:
        if token.text in skill_keywords:
            skills.append(token.text)

    # Extract experience (look for years or job titles)
    for ent in doc.ents:
        if ent.label_ == "DATE" and any(char.isdigit() for char in ent.text):
            experience.append(ent.text)
        if ent.label_ == "ORG":
            experience.append(ent.text)

    # Extract education (look for degrees or institutions)
    for ent in doc.ents:
        if ent.label_ in ["ORG", "NORP"] and any(keyword in ent.text.lower() for keyword in ["university", "college", "degree"]):
            education.append(ent.text)

    return {"skills": list(set(skills)), "experience": list(set(experience)), "education": list(set(education))}

# Function to anonymize resume (remove names, gender, etc.)
def anonymize_text(text):
    doc = nlp(text)
    anonymized_text = text
    for ent in doc.ents:
        if ent.label_ in ["PERSON", "GPE"]:
            anonymized_text = anonymized_text.replace(ent.text, "[REDACTED]")
    anonymized_text = re.sub(r"\b(he|she|him|her|male|female)\b", "[REDACTED]", anonymized_text, flags=re.IGNORECASE)
    return anonymized_text

# Function to get BERT embeddings (FIXED)
def get_bert_embeddings(text):
    inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True, max_length=512)
    inputs = {key: val.to(device) for key, val in inputs.items()}
    with torch.no_grad():
        outputs = model(**inputs)
    embeddings = outputs.last_hidden_state.mean(dim=1).squeeze().detach().cpu().numpy()
    return embeddings

# Function to calculate resume score
def calculate_resume_score(resume_data, job_desc, weights):
    resume_text = " ".join(resume_data["skills"] + resume_data["experience"] + resume_data["education"])
    job_desc_embedding = get_bert_embeddings(job_desc)
    resume_embedding = get_bert_embeddings(resume_text)

    # Calculate cosine similarity
    similarity = cosine_similarity([resume_embedding], [job_desc_embedding])[0][0]

    # Weighted score
    skill_match = len(set(resume_data["skills"]) & set(job_desc.lower().split())) / max(len(job_desc.lower().split()), 1)
    experience_years = len(resume_data["experience"]) / 10  # Simple normalization
    score = (weights["skills"] * skill_match +
             weights["similarity"] * similarity +
             weights["experience"] * experience_years)
    return score, skill_match

# Streamlit app
def main():
    st.title("AI-Based Resume Screening System")
    st.write("Upload resumes and enter job description to screen candidates.")

    # Job description input
    st.header("Job Description")
    job_title = st.text_input("Job Title")
    required_skills = st.text_area("Required Skills (e.g., Python, Data Analysis)")
    required_experience = st.text_area("Required Experience (e.g., 3 years in project management)")
    job_desc = f"{job_title} {required_skills} {required_experience}"

    # Weights for scoring
    st.header("Scoring Weights")
    weight_skills = st.slider("Weight for Skills Match", 0.0, 1.0, 0.5)
    weight_similarity = st.slider("Weight for Semantic Similarity", 0.0, 1.0, 0.3)
    weight_experience = st.slider("Weight for Experience", 0.0, 1.0, 0.2)
    weights = {"skills": weight_skills, "similarity": weight_similarity, "experience": weight_experience}

    # Resume upload
    st.header("Upload Resumes (PDF)")
    uploaded_files = st.file_uploader("Choose PDF resumes", type="pdf", accept_multiple_files=True)

    if st.button("Screen Resumes"):
        if not job_desc or not uploaded_files:
            st.error("Please provide a job description and upload at least one resume.")
            return

        results = []
        embeddings = []

        # Process each resume
        for file in uploaded_files:
            text = extract_text_from_pdf(file)
            anonymized_text = anonymize_text(text)
            resume_data = parse_resume(anonymized_text)
            score, skill_match = calculate_resume_score(resume_data, job_desc, weights)
            results.append({
                "File Name": file.name,
                "Score": score,
                "Skills": ", ".join(resume_data["skills"]),
                "Experience": ", ".join(resume_data["experience"]),
                "Education": ", ".join(resume_data["education"]),
                "Skill Match": f"{skill_match*100:.2f}%"
            })
            embeddings.append(get_bert_embeddings(" ".join(resume_data["skills"] + resume_data["experience"] + resume_data["education"])))

        # Convert results to DataFrame
        df = pd.DataFrame(results)
        df = df.sort_values(by="Score", ascending=False).reset_index(drop=True)

        # Clustering candidates (intelligent feature)
        if len(embeddings) > 1:
            kmeans = KMeans(n_clusters=min(3, len(embeddings)), random_state=42)
            clusters = kmeans.fit_predict(embeddings)
            df["Cluster"] = clusters
            st.write("Candidates grouped into clusters based on similarity:")
            for cluster in range(min(3, len(embeddings))):
                st.write(f"**Cluster {cluster}**: {', '.join(df[df['Cluster'] == cluster]['File Name'])}")

        # Display results
        st.header("Screening Results")
        st.write("Top candidates ranked by relevance to the job description:")
        st.dataframe(df)

        # Download results as CSV
        csv = df.to_csv(index=False).encode('utf-8')
        st.download_button("Download Results as CSV", csv, "screening_results.csv", "text/csv")

if __name__ == "__main__":
    main()
